dash/testing/browser.py

```diff
@@ -191,6 +194,19 @@ def wait_for_page(self, url=None, timeout=10):
                 )
             )
 
+    def toggle_window(self):
+        """switch between the current working window and the new opened one"""
+        if len(self.driver.window_handles) < 2:
+            raise BrowserError("there is no new opening window")
+
+        idx = (self._window_idx + 1) % 2
+        self.switch_window(idx=idx)
+        self._window_idx += 1
+
+    def switch_window(self, idx=0):
+        self.driver.switch_to.window(self.driver.window_handles[idx])
+        wait(0.5)
```

Comment:
`switch_to.window` doesn't incorporate sufficient wait on its own? Or are you waiting for something beyond just having the new window visible?

Reply:
good question, the switch_to.window is a wrapper of `execute` lower level selenium command

``` python
        if self._driver.w3c:
            self._w3c_window(window_name)
            return
        data = {'name': window_name}
        self._driver.execute(Command.SWITCH_TO_WINDOW, data)

    def _w3c_window(self, window_name):
        def send_handle(h):
            self._driver.execute(Command.SWITCH_TO_WINDOW, {'handle': h})

        try:
            # Try using it as a handle first.
            send_handle(window_name)
        except NoSuchWindowException as e:
            # Check every window to try to find the given window name.```

I think it makes sense here to add a wait margin as window-switch is not that responsive, this can avoid having this extra wait everywhere in the test case. 

the  0.5s needs more on field testing, could be 0.1 in our case as dash is really fast, I remember an old project I had I put 2s before.